File: src/main/java/iteration2/Transcript.java
package iteration2;
import com.fasterxml.jackson.annotation.JsonProperty;
import iteration2.StudentSemester;

import java.util.ArrayList;
import java.util.List;

public class Transcript {

    private Float gpa;
    @JsonProperty("semesters")
    private List<StudentSemester> semesters;
    private Integer cumulativeCredit;

    public Transcript () {
        semesters = new ArrayList<StudentSemester>();
    }



    public void calculateGpa() {
        calculateCumulativeCredit();
        float grade = 0;
        for (int i = 0; i < semesters.size(); i++) {
            grade += semesters.get(i).getYano() * semesters.get(i).getCompletedCredit();
        }
        this.gpa =  (float)(((int)((grade / cumulativeCredit) * 100 )) / 100.0);
    }

    public void calculateCumulativeCredit() {
        cumulativeCredit = 0;
        for (int i = 0; i < semesters.size(); i++) {
            cumulativeCredit += semesters.get(i).getCompletedCredit();
        }
    }

    public Float getGpa() {
        return gpa;
    }

    public void setGpa(Float gpa) {
        this.gpa = gpa;
    }

    public List<StudentSemester> getSemesters() {
        return semesters;
    }

    public void setSemesters(List<StudentSemester> semesters) {
        this.semesters = semesters;
    }

    public boolean containsCourse(String courseCode){
        for (StudentSemester semester : semesters){
            for (GivenCourse course : semester.getGivenCourses()){
                if (course.getCourseCode().equals(courseCode)){
                    return true;
                }
            }
        }
        return false;
    }
    public GivenCourse getGivenCourse(String courseCode){
        for (StudentSemester semester : semesters){
            for (GivenCourse course : semester.getGivenCourses()){
                if (course.getCourseCode().equals(courseCode)){
                    return course;
                }
            }
        }
        return null;
    }

    public float getMaxGrade(String courseCode){
        float maxGrade = 0;
        for (StudentSemester semester : semesters){
            for (GivenCourse course : semester.getGivenCourses()){
                if (course.getCourseCode().equals(courseCode) && course.getGrade() > maxGrade){
                    maxGrade = course.getGrade();
                }
            }
        }
        return maxGrade;
    }

    @Override
    public String toString() {
        return "Transcript{" +
                "gpa=" + gpa +
                ", semesters=" + semesters +
                ", cumulativeCredit=" + cumulativeCredit +
                '}';
    }
}

File: Transcript.py
from typing import List

from CSE3063F22P1_GRP4.StudentSemester import StudentSemester


class Transcript:
    def __init__(self):
        self.gpa = None
        self.semesters = []
        self.cumulative_credit = 0

    def calculate_gpa(self):
        self.calculate_cumulative_credit()
        grade = 0
        for i in range(len(self.semesters)):
            grade += self.semesters[i].yano * self.semesters[i].completed_credit
        self.gpa = (int((grade / self.cumulative_credit) * 100)) / 100.0

    def calculate_cumulative_credit(self):
        self.cumulative_credit = 0
        for semester in self.semesters:
            self.cumulative_credit += semester.completed_credit

    def get_gpa(self):
        return self.gpa

    def set_gpa(self, gpa):
        self.gpa = gpa

    def get_semesters(self) -> List[StudentSemester]:
        return self.semesters

    def set_semesters(self, semesters):
        self.semesters = semesters

    def contains_course(self, course_code):
        for semester in self.semesters:
            for course in semester.given_courses:
                if course.courseCode == course_code:
                    return True
        return False

    def get_given_course(self, course_code):
        for semester in self.semesters:
            for course in semester.get_given_courses():
                if course.get_course_code() == course_code:
                    return course
        return None

    def get_max_grade(self, course_code):
        max_grade = 0
        for semester in self.semesters:
            for course in semester.given_courses:
                if course.courseCode == course_code and course.grade > max_grade:
                    max_grade = course.grade
        return max_grade

    def __repr__(self):
        return f"Transcript(gpa={self.gpa}, semesters={self.semesters}, cumulative_credit={self.cumulative_credit})"

